### PR TITLE
Bug 1626258 - Better handling of global object in a threadsafe manner.

### PR DESCRIPTION
**Pull request by badboy**
This has two things:

* Getting the global object now gives an option and thus the FFI layer can return an error instead of panicking.
* Better docs on why the other code is ok, with the addition of logging an error instead of panicking.

Again: it is unlikely to ever hit any of these cases for our current wrappers, as they only initialize once.

### CODE REVIEW

**Comment by badboy** [comment]
(I'd like two pairs of eyes on this, just to be on the safe side)

File: glean-core/src/lib.rs [hunk 1]
```diff
@@ -81,15 +81,31 @@ static GLEAN: OnceCell<Mutex<Glean>> = OnceCell::new();
 /// Get a reference to the global Glean object.
 ///
 /// Panics if no global Glean object was set.
```
Comment (travis79) [inline glean-core/src/lib.rs:83]: Is this comment still true after the protections you added?

Comment (badboy) [reply glean-core/src/lib.rs:83]: Of course not, missed that. thanks!

File: glean-core/src/error.rs [hunk 2]
```diff
@@ -54,6 +54,9 @@ pub enum ErrorKind {
     /// Unknown error
     Utf8Error,
 
+    /// Glean not initialized
+    Missing,
```
Comment (Dexterp37) [inline glean-core/src/error.rs:58]: nit: what about `NotInitialized`?

Comment (badboy) [reply glean-core/src/error.rs:58]: Better wording indeed.

**Comment by badboy** [comment]
> I wonder if this will have any impact on the obscure errors we're seeing on Android through sentry, where kotlin is attempting to call functions on uninitialized objects. Maybe this might help bisect, if the problem happens because of Rust panicking?

So far these calls have been on the Kotlin object being null though.

**Comment by Dexterp37** [comment]
> 
> 
> > I wonder if this will have any impact on the obscure errors we're seeing on Android through sentry, where kotlin is attempting to call functions on uninitialized objects. Maybe this might help bisect, if the problem happens because of Rust panicking?
> 
> So far these calls have been on the Kotlin object being null though.

Yes, but what if the object was null because, for some reason, Rust panicked and Kotlin failed to "init the object"? I mean, I have no idea on how this would even be possible, but maybe it is :D

**Pull request merged**